In [3]:
import pandas as pd

## Условие 1: 1 задача
Скачать файл в уроке \
Считать данные с помощью pandas  
Вывести на экран первые 5 строк  
Посмотреть на описание признаков и на их содержание

In [5]:
df = pd.read_csv('kc_final.csv')
df

Unnamed: 0          id        date     price  bedrooms  bathrooms  \
0               0  7129300520  2014-10-13  221900.0         3       1.00   
1               1  6414100192  2014-12-09  538000.0         3       2.25   
2               2  5631500400  2015-02-25  180000.0         2       1.00   
3               3  2487200875  2014-12-09  604000.0         4       3.00   
4               4  1954400510  2015-02-18  510000.0         3       2.00   
...           ...         ...         ...       ...       ...        ...   
21608       21608   263000018  2014-05-21  360000.0         3       2.50   
21609       21609  6600060120  2015-02-23  400000.0         4       2.50   
21610       21610  1523300141  2014-06-23  402101.0         2       0.75   
21611       21611   291310100  2015-01-16  400000.0         3       2.50   
21612       21612  1523300157  2014-10-15  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  ...  grade  sqft_above  \
0             1180      5650     1.0           0  ...      7        1180   
1             2570      7242     2.0           0  ...      7        2170   
2              770     10000     1.0           0  ...      6         770   
3             1960      5000     1.0           0  ...      7        1050   
4             1680      8080     1.0           0  ...      8        1680   
...            ...       ...     ...         ...  ...    ...         ...   
21608         1530      1131     3.0           0  ...      8        1530   
21609         2310      5813     2.0           0  ...      8        2310   
21610         1020      1350     2.0           0  ...      7        1020   
21611         1600      2388     2.0           0  ...      8        1600   
21612         1020      1076     2.0           0  ...      7        1020   

       sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0                  0      1955             0    98178  47.5112 -122.257   
1                400      1951          1991    98125  47.7210 -122.319   
2                  0      1933             0    98028  47.7379 -122.233   
3                910      1965             0    98136  47.5208 -122.393   
4                  0      1987             0    98074  47.6168 -122.045   
...              ...       ...           ...      ...      ...      ...   
21608              0      2009             0    98103  47.6993 -122.346   
21609              0      2014             0    98146  47.5107 -122.362   
21610              0      2009             0    98144  47.5944 -122.299   
21611              0      2004             0    98027  47.5345 -122.069   
21612              0      2008             0    98144  47.5941 -122.299   

       sqft_living15  sqft_lot15  
0               1340        5650  
1               1690        7639  
2               2720        8062  
3               1360        5000  
4               1800        7503  
...              ...         ...  
21608           1530        1509  
21609           1830        7200  
21610           1020        2007  
21611           1410        1287  
21612           1020        1357  

[21613 rows x 22 columns]

In [3]:
df.head()

Unnamed: 0          id        date     price  bedrooms  bathrooms  \
0           0  7129300520  2014-10-13  221900.0         3       1.00   
1           1  6414100192  2014-12-09  538000.0         3       2.25   
2           2  5631500400  2015-02-25  180000.0         2       1.00   
3           3  2487200875  2014-12-09  604000.0         4       3.00   
4           4  1954400510  2015-02-18  510000.0         3       2.00   

   sqft_living  sqft_lot  floors  waterfront  ...  grade  sqft_above  \
0         1180      5650     1.0           0  ...      7        1180   
1         2570      7242     2.0           0  ...      7        2170   
2          770     10000     1.0           0  ...      6         770   
3         1960      5000     1.0           0  ...      7        1050   
4         1680      8080     1.0           0  ...      8        1680   

   sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0              0      1955             0    98178  47.5112 -122.257   
1            400      1951          1991    98125  47.7210 -122.319   
2              0      1933             0    98028  47.7379 -122.233   
3            910      1965             0    98136  47.5208 -122.393   
4              0      1987             0    98074  47.6168 -122.045   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062  
3           1360        5000  
4           1800        7503  

[5 rows x 22 columns]

## Условие 2: 2 задача
Проведите первичный анализ данных  
Изучите типы данных  
Найдите количество пропущенных ячеек в данных  
Посчитайте основные статистики по всем признакам и поизучайте их  
Пишите выводы   

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 22 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Unnamed: 0     21613 non-null  int64  
 1   id             21613 non-null  int64  
 2   date           21613 non-null  object 
 3   price          21613 non-null  float64
 4   bedrooms       21613 non-null  int64  
 5   bathrooms      21613 non-null  float64
 6   sqft_living    21613 non-null  int64  
 7   sqft_lot       21613 non-null  int64  
 8   floors         21613 non-null  float64
 9   waterfront     21613 non-null  int64  
 10  view           21613 non-null  int64  
 11  condition      21613 non-null  int64  
 12  grade          21613 non-null  int64  
 13  sqft_above     21613 non-null  int64  
 14  sqft_basement  21613 non-null  int64  
 15  yr_built       21613 non-null  int64  
 16  yr_renovated   21613 non-null  int64  
 17  zipcode        21613 non-null  int64  
 18  lat   

In [6]:
df.dtypes

Unnamed: 0         int64
id                 int64
date              object
price            float64
bedrooms           int64
bathrooms        float64
sqft_living        int64
sqft_lot           int64
floors           float64
waterfront         int64
view               int64
condition          int64
grade              int64
sqft_above         int64
sqft_basement      int64
yr_built           int64
yr_renovated       int64
zipcode            int64
lat              float64
long             float64
sqft_living15      int64
sqft_lot15         int64
dtype: object

In [7]:
df.isnull().sum()

Unnamed: 0       0
id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

In [9]:
df.describe()

Unnamed: 0            id         price      bedrooms     bathrooms  \
count  21613.00000  2.161300e+04  2.161300e+04  21613.000000  21613.000000   
mean   10806.00000  4.580302e+09  5.400881e+05      3.370842      2.114757   
std     6239.28002  2.876566e+09  3.671272e+05      0.930062      0.770163   
min        0.00000  1.000102e+06  7.500000e+04      0.000000      0.000000   
25%     5403.00000  2.123049e+09  3.219500e+05      3.000000      1.750000   
50%    10806.00000  3.904930e+09  4.500000e+05      3.000000      2.250000   
75%    16209.00000  7.308900e+09  6.450000e+05      4.000000      2.500000   
max    21612.00000  9.900000e+09  7.700000e+06     33.000000      8.000000   

        sqft_living      sqft_lot        floors    waterfront          view  \
count  21613.000000  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean    2079.899736  1.510697e+04      1.494309      0.007542      0.234303   
std      918.440897  4.142051e+04      0.539989      0.086517      0.766318   
min      290.000000  5.200000e+02      1.000000      0.000000      0.000000   
25%     1427.000000  5.040000e+03      1.000000      0.000000      0.000000   
50%     1910.000000  7.618000e+03      1.500000      0.000000      0.000000   
75%     2550.000000  1.068800e+04      2.000000      0.000000      0.000000   
max    13540.000000  1.651359e+06      3.500000      1.000000      4.000000   

       ...         grade    sqft_above  sqft_basement      yr_built  \
count  ...  21613.000000  21613.000000   21613.000000  21613.000000   
mean   ...      7.656873   1788.390691     291.509045   1971.005136   
std    ...      1.175459    828.090978     442.575043     29.373411   
min    ...      1.000000    290.000000       0.000000   1900.000000   
25%    ...      7.000000   1190.000000       0.000000   1951.000000   
50%    ...      7.000000   1560.000000       0.000000   1975.000000   
75%    ...      8.000000   2210.000000     560.000000   1997.000000   
max    ...     13.000000   9410.000000    4820.000000   2015.000000   

       yr_renovated       zipcode           lat          long  sqft_living15  \
count  21613.000000  21613.000000  21613.000000  21613.000000   21613.000000   
mean      84.402258  98077.939805     47.560053   -122.213896    1986.552492   
std      401.679240     53.505026      0.138564      0.140828     685.391304   
min        0.000000  98001.000000     47.155900   -122.519000     399.000000   
25%        0.000000  98033.000000     47.471000   -122.328000    1490.000000   
50%        0.000000  98065.000000     47.571800   -122.230000    1840.000000   
75%        0.000000  98118.000000     47.678000   -122.125000    2360.000000   
max     2015.000000  98199.000000     47.777600   -121.315000    6210.000000   

          sqft_lot15  
count   21613.000000  
mean    12768.455652  
std     27304.179631  
min       651.000000  
25%      5100.000000  
50%      7620.000000  
75%     10083.000000  
max    871200.000000  

[8 rows x 21 columns]

## Условие 3: 3 задача
Ответьте на несколько вопросов  
### 3.1 В каком диапазоне изменяются стоимости недвижимости?  


In [12]:
df['price'].min()
df['price'].max()
print(f"Диапазон изменения стоимости недвижимости: {df['price'].min()}$ - {df['price'].max()}$")

Диапазон изменения стоимости недвижимости: 75000.0$ - 7700000.0$


### 3.2 Какую долю в среднем занимают жилая площадь от всей площади по всем домам?  


In [18]:
df['sqft_living'].sum()

44952873

In [19]:
df['sqft_lot'].sum()

326506890

In [20]:
print(f"Доля жилой площади от общей: {(df['sqft_living'].sum() * 100)/df['sqft_lot'].sum()} ")

Доля жилой площади от общей: 13.767817579592272 


### 3.3 Как много домов с разными этажами в данных?  
 

In [23]:
df['floors'].unique()

array([1. , 2. , 1.5, 3. , 2.5, 3.5])

In [24]:
df['floors'].nunique()

6

В данных представлены дома с 6 разными этажами

### 3.4 Насколько хорошие состояния у домов в данных?  


In [7]:
df['condition'].mean()

3.4094295100171195

In [8]:
print(f"Минимальная оценка состаяния домов: {df['condition'].min()}, максимальная - {df['condition'].max()}, средняя - {df['condition'].mean()}")

Минимальная оценка состаяния домов: 1, максимальная - 5, средняя - 3.4094295100171195


Состояние домов в среднем - нормальное


### 3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных? 

In [27]:
df['yr_built'].iloc[[0, 21612]]

0        1955
21612    2008
Name: yr_built, dtype: int64


## Условие 4: 4 задача
Ответьте на несколько вопросов

### 4.1 Сколько в среднем стоят дома, у которых 2 спальни?


In [28]:
df[df['bedrooms'] == 2]['price'].mean()

401372.681884058

### 4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?


In [6]:
df[df['price'] > 600000.0]['sqft_lot'].mean()

20442.524776214832

### 4.3 Как много домов коснулся ремонт?



Дома с ремонтом:

In [15]:
df[(df['yr_renovated'] != 0)].shape[0]

914

Домов всего:

In [16]:
df['yr_renovated'].shape[0]

21613

Ответ:

In [17]:
(df[(df['yr_renovated'] != 0)].shape[0] * 100)/df['yr_renovated'].shape[0]

4.228936288344977

Всего в 4% домов был сделан ремонт

### 4.4 На сколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?

In [25]:
good_fl = df[df['grade'] > 10]['price'].mean()
good_fl

1678635.1175298805

In [26]:
bad_fl = df[df['grade'] < 4]['price'].mean()
bad_fl

189750.0

In [22]:
df[df['grade'] > 10]['price'].mean() - df[df['grade'] < 4]['price'].mean()

1488885.1175298805

In [33]:
((good_fl * 100)/bad_fl)

884.6561884215444

**Ответ:** квартиры с хорошим рейтингом стоят в среднем на 1488885$ или на 884.65% дороже

## Условие 5: 5 задача
Ответьте на несколько вопросов


### 5.1 Выберите дом клиенту
Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?


In [36]:
df[df['waterfront'] == 1][df['bathrooms'] > 3][df['sqft_basement'] != 0].shape[0]

<ipython-input-36-bce9a8776b86>:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df[df['waterfront'] == 1][df['bathrooms'] > 3][df['sqft_basement'] != 0].shape[0]


36

### 5.2 Выберите дом клиенту
Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В какой ценовом диапазоне будут дома?


In [43]:
choiсe = df[((df['view'] >= 3) | (df['waterfront'] != 1)) & (df['yr_built'] >= 1980) & (df['condition'] == 5)]['price']
print(f"Ценовой диапазон по данным параметрам: от {choiсe.min()} до {choiсe.max()}")

Ценовой диапазон по данным параметрам: от 205000.0 до 3000000.0


### 5.3 Выберите дом клиенту
Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?

In [44]:
df[(df['sqft_basement'] == 0) & (df['floors'] == 2) & (df['price'] < 150000)]['condition'].mean()

2.8333333333333335